### PR TITLE
Add `enabled_if` to `install` stanzas and enforce variable restrictions

### LINKED ISSUE #3354: Allow to put `enabled_if` in install stanza
It would be nice if we could control the `install` stanza a bit more, for instance by deciding if something has to be installed using the `%{ocaml-config:system}` variable.

But it seems that it's not possible to put an `enabled_if` condition inside `install` stanza. Would it be possible to add it?

### PR DESCRIPTION
**Pull request by voodoos**
This PR fixes #3354 by adding the `enabled_if` field to the `install` stanza.

Doing that I discovered that the `enabled_if` field was not guarded correctly in the executable stanza and any variables could be used. This PR enforce the same variable restrictions for `enabled_if` in the executables and install stanzas as in the library stanza.

To do that chose to upstream the ad-hoc variable checking code for the install stanza to the enabled_if decoder.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Needs a CHANGES entry.

There's a test failure as well:

```
--- a/test/blackbox-tests/test-cases/enabled_if/run.t
463+++ b/test/blackbox-tests/test-cases/enabled_if/run.t.corrected
464@@ -24,8 +24,8 @@ Test the enabled_if field for libraries:
465   [1]
466 
467   $ dune build main.exe
468-  File "dune", line 34, characters 12-15:
469-  34 |  (libraries foo))
470+  File "dune", line 35, characters 12-15:
471+  35 |  (libraries foo))
472                    ^^^
473   Error: Library "foo" in _build/default is hidden (unsatisfied 'enabled_if').
```

**Comment by voodoos** [comment]
> Needs a CHANGES entry.
> 
> There's a test failure as well

Done.

File: src/dune/dune_file.ml [hunk 1]
```diff
@@ -1053,14 +1035,32 @@ module Install_conf = struct
...
        and+ files = field "files" File_binding.Unexpanded.L.decode
-       and+ package = Pkg.field "install" in
-       { section; files; package })
+       and+ package = Pkg.field "install"
+       and+ enabled_if =
+         let allowed_vars =
+           Some
+             (List.map
+                ~f:(function
+                  | var -> (var, (2, 6)))
```
Comment (rgrinberg) [inline src/dune/dune_file.ml:1051]: Why not just `fun var ->`?

File: src/dune/dune_file.ml [hunk 2]
```diff
@@ -1486,7 +1487,23 @@ module Executables = struct
...
+          (List.map
+             ~f:(function
+               | var -> (var, (2, 6)))
+             [ "architecture"
+             ; "system"
+             ; "model"
+             ; "os_type"
+             ; "ccomp_type"
+             ; "profile"
+             ; "ocaml_version"
```
Comment (rgrinberg) [inline src/dune/dune_file.ml:1502]: Hmm, the duplication between this list and the one above is a bit unfortunate. Can we share the common elements?

File: src/dune/dune_file.ml (hunk 1 above)
Comment (voodoos) [reply src/dune/dune_file.ml:1051]: It was to stress the fact that other variables can be assigned different minimal version.
But I will factorize it with the list.

**Comment by voodoos** [comment]
I decided to move the `enabled_if` related code to it's own module to stop bloating the `Dune_file` module.

It is not completely generic yet and does not replace the code in `Lib_info` that associates different minimal version to each variable. (https://github.com/ocaml/dune/blob/07825785dddb38cfced908ea5b4e9f21c1f19c4f/src/dune/lib_config.ml#L29)

However I feel like it could be easily added when needed (when another variable will be allowed in these fields).

What do you think ?

File: test/blackbox-tests/test-cases/enabled_if-exec/install/dune [hunk 3]
```diff
@@ -0,0 +1,15 @@
+(dirs :standard \ forbidden_var)
+
+(executable
+  (name foo)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/enabled_if-exec/install/dune:4]: This is a bit oddly indented. Should be 1 space like the rest of the code.

**Comment by rgrinberg** [comment]
The `Enabled_if` module looks good to me.

**Pull request merged**